docs/proposals/api-group.md

```diff
@@ -0,0 +1,150 @@
+
+    For types like GroupAndVersionList, VersionList, APIResourceList (these will be used by API server to return the supported group/version, version, and resources during API discovery), we plan to take the latter approach by placing them in a new package called `unversioned`. Types in `unversioned` package will not have the TypeMeta field.
+
+  * non-top-level common API objects:
+
+    Assuming object o belonging to group X is used as a field in an object belonging to group Y, currently genconversion will generate the conversion functions for o in package Y. Hence, we don't need any special treatment for non-top-level common API objects.
+
+    TypeMeta is an exception, because it is a common object that is used by objects in all groups but does not logically belong to any group. We plan to move it to the package `unversioned`.
+
+## Client-side implementation
+
+1. clients:
+
+  Currently we have structured (pkg/client/unversioned#ExperimentalClient, pkg/client/unversioned#Client) and unstructured (pkg/kubectl/resource#Helper) clients. The structured clients are not scalable because each of them implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). Only the unstructured clients are scalable. We should either auto-generate the code for structured clients or migrate to use the unstructured clients as much as possible.
+
+2. Spelling the URL:
+
+  The URL is in the form of prefix/group/version/. The prefix is hard-coded in the client/unversioned.Config (see [here](../../pkg/client/unversioned/experimental.go#L101)). The client should be able to figure out `group` and `version` using the RESTMapper. For a third-party client which does not have access to the RESTMapper, it should discover the mapping of `group`, `version` and `kind` by querying the server as described in point 2 of #server-side-implementation.
+
+3. kubectl:
+
+  kubectl should accept arguments like `group/resource`, `group/resource/name`. Nevertheless, the user can omit the `group`, then kubectl shall rely on RESTMapper.VersionAndKindForResource() to figure out which group the resource belongs to, or return an error if there is ambiguity. A special case is when a resource (like `node`) exist in both k8s v1 API and k8s modularized API (like `infra/v2`), kubectl should use the modularized API rather than returning an error because of the ambiguity.
+
+  The --api-version flag of kubectl should accept values in the form of `group/version`.
```

Comment:
We probably don't need it yet, but should reexamine it when we talk about
discovery (if we record discovery we may not need to store preferred
version the same way)

On Sep 2, 2015, at 9:19 PM, Chao Xu <notifications@github.com> wrote:

In docs/proposals/api-group.md
<https://github.com/kubernetes/kubernetes/pull/12951#discussion_r38604849>:

> +
> +## Client-side implementation
> +
> +1. clients:
> +
> +  Currently we have structured (pkg/client/unversioned#ExperimentalClient, pkg/client/unversioned#Client) and unstructured (pkg/kubectl/resource#Helper) clients. The structured clients are not scalable because each of them implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). Only the unstructured clients are scalable. We should either auto-generate the code for structured clients or migrate to use the unstructured clients as much as possible.
> +
> +2. Spelling the URL:
> +
> +  The URL is in the form of prefix/group/version/. The prefix is hard-coded in the client/unversioned.Config (see [here](../../pkg/client/unversioned/experimental.go#L101)). The client should be able to figure out `group` and `version` using the RESTMapper. For a third-party client which does not have access to the RESTMapper, it should discover the mapping of `group`, `version` and `kind` by querying the server as described in point 2 of #server-side-implementation.
> +
> +3. kubectl:
> +
> +  kubectl should accept arguments like `group/resource`, `group/resource/name`. Nevertheless, the user can omit the `group`, then kubectl shall rely on RESTMapper.VersionAndKindForResource() to figure out which group the resource belongs to, or return an error if there is ambiguity. A special case is when a resource (like `node`) exist in both k8s v1 API and k8s modularized API (like `infra/v2`), kubectl should use the modularized API rather than returning an error because of the ambiguity.
> +
> +  The --api-version flag of kubectl should accept values in the form of `group/version`.

I agree the global apiVersion setting shoudl be deprecated. Shall we add a
groupVersion []string field to allow the user to configure the preferred
version of each group?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12951/files#r38604849>.